From: IrfanMohammad <irfan@graphiant.com>
Date: Tue, 22 Aug 2023 11:59:21 +0000
Subject: Verify data parse before store

If due to any reason, bad data is written into a datastore, it is not
recoverable until we reboot the system. In that case it is better to not
have written it.

Dump the failed datastore file to .failed file for debugging
---
 src/plugins/ds_json.c | 89 ++++++++++++++++++++++++++++++++++++++++++++++-----
 1 file changed, 81 insertions(+), 8 deletions(-)

diff --git a/src/plugins/ds_json.c b/src/plugins/ds_json.c
index 14f9834..ac2dbac 100644
--- a/src/plugins/ds_json.c
+++ b/src/plugins/ds_json.c
@@ -45,16 +45,70 @@ static sr_error_info_t *srpds_json_load(const struct lys_module *mod, sr_datasto
 static sr_error_info_t * srpds_json_access_get(const struct lys_module *mod, sr_datastore_t ds, void *plg_data,
         char **owner, char **group, mode_t *perm);
 
+static void
+srpds_json_store_failed(struct iovec *iov, const char *path)
+{
+    int fd;
+    char *fail_path = NULL;
+    int ret = asprintf(&fail_path, "%s.failed", path);
+
+    SRPLG_LOG_INF(srpds_name, "Dumping failed for %s", path);
+    if (ret < 0) {
+        return;
+    }
+
+    SRPLG_LOG_INF(srpds_name, "Dumping to %s", fail_path);
+    fd = srpjson_open(srpds_name, fail_path, O_CREAT | O_WRONLY, 00600);
+    if (fd < 0) {
+        goto cleanup;
+    }
+    SRPLG_LOG_INF(srpds_name, "Opened %s", fail_path);
+
+    if (ftruncate(fd, 0) < 0) {
+        goto cleanup;
+    }
+
+    SRPLG_LOG_INF(srpds_name, "Truncated %s", fail_path);
+    if (srpjson_writev(srpds_name, fd, iov, 1)) {
+        goto cleanup;
+    }
+    SRPLG_LOG_INF(srpds_name, "Written %s", fail_path);
+
+cleanup:
+    if (fd > 0) {
+        close(fd);
+    }
+    free(fail_path);
+}
+
 static sr_error_info_t *
 srpds_json_store_(const struct lys_module *mod, sr_datastore_t ds, const struct lyd_node *mod_data, const char *owner,
         const char *group, mode_t perm, int make_backup)
 {
+    int rc;
     sr_error_info_t *err_info = NULL;
     struct stat st;
     char *path = NULL, *bck_path = NULL;
     int fd = -1, backup = 0, creat = 0;
-    uint32_t print_opts;
-    off_t size;
+    uint32_t print_opts, parse_opts;
+    struct iovec iov;
+
+    struct lyd_node *read_back = NULL;
+    char *str = NULL;
+
+    /* set parse options */
+    parse_opts = LYD_PARSE_ONLY | LYD_PARSE_ORDERED;
+    if (ds == SR_DS_OPERATIONAL) {
+      /* edit may include opaque nodes */
+      parse_opts |= LYD_PARSE_OPAQ;
+    } else {
+      parse_opts |= LYD_PARSE_STRICT;
+    }
+    if ((ds == SR_DS_RUNNING) || (ds == SR_DS_STARTUP) || (ds == SR_DS_FACTORY_DEFAULT)) {
+      /* always valid datastores */
+      parse_opts |= LYD_PARSE_WHEN_TRUE | LYD_PARSE_NO_NEW;
+    }
+
 
     /* get path */
     if ((err_info = srpjson_get_path(srpds_name, mod->name, ds, &path))) {
@@ -124,24 +178,43 @@ srpds_json_store_(const struct lys_module *mod, sr_datastore_t ds, const struct
 
     /* print data */
     print_opts = LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS | LYD_PRINT_KEEPEMPTYCONT | LYD_PRINT_WD_IMPL_TAG;
-    if (lyd_print_fd(fd, mod_data, LYD_JSON, print_opts)) {
+
+    if (lyd_print_mem_len(&str, mod_data, LYD_JSON, print_opts, &iov.iov_len)) {
         err_info = srpjson_log_err_ly(srpds_name, LYD_CTX(mod_data));
         srplg_log_errinfo(&err_info, srpds_name, NULL, SR_ERR_INTERNAL, "Failed to store data into \"%s\".", path);
         goto cleanup;
     }
 
-    /* truncate the file to the exact size (to get rid of possible following old data) */
-    if ((size = lseek(fd, 0, SEEK_CUR)) == -1) {
+    iov.iov_base = str;
+
+    if (mod_data) {
+        rc = lyd_parse_data_mem(LYD_CTX(mod_data), str, LYD_JSON, parse_opts, 0, &read_back);
+        lyd_free_all(read_back);
+        if (rc) {
+            err_info = srpjson_log_err_ly(srpds_name, LYD_CTX(mod_data));
+            srplg_log_errinfo(&err_info, srpds_name, NULL, SR_ERR_INTERNAL, "Failed to read back data to be stored for mod %s", mod->name);
+
+            /* Dump it to logs */
+            srpds_json_store_failed(&iov, path);
+            goto cleanup;
+        }
+    }
+
+    if (ftruncate(fd, 0) == -1) {
         srplg_log_errinfo(&err_info, srpds_name, NULL, SR_ERR_SYS, "Failed to get the size of \"%s\" (%s).", path,
                 strerror(errno));
         goto cleanup;
     }
-    if (ftruncate(fd, size) == -1) {
-        srplg_log_errinfo(&err_info, srpds_name, NULL, SR_ERR_SYS, "Failed to truncate \"%s\" (%s).", path,
-                strerror(errno));
+
+    /* dump str to fd */
+    if (srpjson_writev(srpds_name, fd, &iov, 1)) {
+        err_info = srpjson_log_err_ly(srpds_name, LYD_CTX(mod_data));
+        srplg_log_errinfo(&err_info, srpds_name, NULL, SR_ERR_INTERNAL, "Failed to store data into \"%s\".", path);
         goto cleanup;
     }
 
+    free(str);
+
 cleanup:
     /* delete the backup file */
     if (backup && (unlink(bck_path) == -1)) {
